Install necessary packages


In [13]:
%pip install pandas
%pip install glob2


[notice] A new release of pip is available: 23.3.2 -> 25.1.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.

[notice] A new release of pip is available: 23.3.2 -> 25.1.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas as pd
import glob

Go do some stuff


In [4]:
# Point to data location
path = "./scraper/data/6_23_2025/"
all_files = glob.glob(path + "*.json")

# Read all files into single dataframe

df_list = []
for file in all_files:
    df = pd.read_json(file)
    df_list.append(df)

combined_df = pd.concat(df_list, ignore_index=True) # ignore_index resets index of combined DF to prevent index duplicates

In [16]:
# Display entire data frame
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

# Display first 10 rows
combined_df.head(5)

stockNumber                vin  year        make              model  \
0     27461478  5NPEG4JA8NH134099  2022     Hyundai             Sonata   
1     26070459  JTDEPMAE5NJ210260  2022      Toyota            Corolla   
2     27175122  3GCUDFE80RG128481  2024   Chevrolet     Silverado 1500   
3     27072297  1V2WE2CA5PC229115  2023  Volkswagen  Atlas Cross Sport   
4     27386026  58AGZ1B12MU110647  2021       Lexus             ES 350   

               body           trim  basePrice  originalPrice  hasPriceDrop  \
0          4D Sedan             SE      19998            NaN         False   
1          4D Sedan             LE      20998            NaN         False   
2       4D Crew Cab  LT Trail Boss      51998            NaN         False   
3  4D Sport Utility      SE w/Tech      32998            NaN         False   
4          4D Sedan        F-Sport      35998            NaN         False   

   mileage  storeId storeName storeCity    state stateAbbreviation  distance  \
0    56782     7298   Roswell   Roswell  Georgia                GA    2108.0   
1    33730     7298   Roswell   Roswell  Georgia                GA    2108.0   
2    58240     7298   Roswell   Roswell  Georgia                GA    2108.0   
3    15895     7298   Roswell   Roswell  Georgia                GA    2108.0   
4    26900     7298   Roswell   Roswell  Georgia                GA    2108.0   

   averageRating  numberOfReviews  repairPalData  isNewArrival  \
0       4.200000                5            NaN         False   
1       4.500000               22            NaN         False   
2       5.000000                1            NaN         False   
3       5.000000                2            NaN         False   
4       4.333333                6            NaN         False   

   isTransferable                                           features  \
0            True  [Rear Defroster, Overhead Airbags, Auxiliary A...   
1            True  [Power Windows, Power Locks, Air Conditioning,...   
2            True  [Power Locks, Auxiliary Audio Input, AM/FM Ste...   
3            True  [Alloy Wheels, 20 Inch Plus Wheels, Power Lock...   
4            True  [Power Windows, Traction Control, ABS Brakes, ...   

                                 highlightedFeatures  \
0  Rear View Camera,Auxiliary Audio Input,Alloy W...   
1  Rear View Camera,Auxiliary Audio Input,Overhea...   
2  Z71 Package,4WD/AWD,Leather Seats,BOSE Sound S...   
3  Technology Package,Leatherette Seats,Satellite...   
4  Leatherette Seats,Pioneer Sound System,Parking...   

                                          highlights  \
0  [singleOwner, warranty, fuelEfficient, advance...   
1        [warranty, fuelEfficient, advancedFeatures]   
2      [singleOwner, advancedFeatures, premiumAudio]   
3  [singleOwner, lowMiles, warranty, advancedFeat...   
4  [singleOwner, lowMilesPerYear, warranty, advan...   

       lastMadeSaleableDate  transferFee            transferTags  \
0  2025-06-12T20:56:13.565Z       1999.0  [GeneratedByCps, Paid]   
1  2025-06-12T20:53:33.311Z       1999.0  [GeneratedByCps, Paid]   
2  2025-06-20T21:39:32.178Z       1999.0  [GeneratedByCps, Paid]   
3  2025-05-28T20:50:22.353Z       1999.0  [GeneratedByCps, Paid]   
4   2025-06-13T20:59:37.45Z       1999.0  [GeneratedByCps, Paid]   

     transferText   transferType  minEstimatedTransferDurationInDays  \
0  $1999 Shipping  shipping-from                                 9.0   
1  $1999 Shipping  shipping-from                                 9.0   
2  $1999 Shipping  shipping-from                                 9.0   
3  $1999 Shipping  shipping-from                                 9.0   
4  $1999 Shipping  shipping-from                                 9.0   

   maxEstimatedTransferDurationInDays  transferTimesUnavailable exteriorColor  \
0                                22.0                     False          Gray   
1                                22.0                     False          Blue   
2                  

Need to dive deeper into the possible values for each column

Data Cleaning
    - handle missing values
    - handle outliers
    
Data Transformation 
    - scaling
    - normalization
    - encoding 
    
Data Splitting
    - create training and testing sets

In [23]:
print("Min ", combined_df['transferFee'].min())
print("Median ",combined_df['transferFee'].median())
print("Mode ",combined_df['transferFee'].mode())

Min  0.0
Median  1999.0
Mode  0    1999.0
Name: transferFee, dtype: float64


In [24]:
# Check data type of features
combined_df.dtypes

stockNumber                int64
vin                       object
year                       int64
make                      object
model                     object
                          ...   
series                    object
isSavedCar                  bool
heroImageUrl              object
heroThumbnailImageUrl    float64
isEVTaxCreditEligible       bool
Length: 69, dtype: object

In [25]:
# Key statistics of studied variables
combined_df.describe()

stockNumber          year     basePrice  originalPrice        mileage  \
count  8.711800e+04  87118.000000  87118.000000    6000.000000   87118.000000   
mean   2.717417e+07   2020.133015  27154.264951   31170.838833   47292.752818   
std    3.649709e+05      2.952196  10909.109501   12141.425771   29595.639905   
min    1.386364e+07   2001.000000   3065.000000    8998.000000      16.000000   
25%    2.704489e+07   2018.000000  19998.000000   21998.000000   24536.500000   
50%    2.726413e+07   2021.000000  24998.000000   28998.000000   42224.500000   
75%    2.742127e+07   2022.000000  31998.000000   37998.000000   64124.750000   
max    2.765300e+07   2025.000000  99998.000000   94998.000000  756951.000000   

            storeId      distance  averageRating  numberOfReviews  \
count  87118.000000  87118.000000   87118.000000     87118.000000   
mean    6836.857286   1603.210494       4.017579        14.774398   
std      544.944284    815.575405       1.450467        18.221099   
min     6001.000000      0.000000       0.000000         0.000000   
25%     6110.000000    926.800000       4.129032         3.000000   
50%     7128.000000   1825.500000       4.500000         9.000000   
75%     7209.000000   2342.800000       4.764706        20.000000   
max     7810.000000   2702.300000       5.000000       202.000000   

       repairPalData   transferFee  minEstimatedTransferDurationInDays  \
count            0.0  51299.000000                        51299.000000   
mean             NaN   1554.085031                            7.996023   
std              NaN    729.866516                            2.753762   
min              NaN      0.000000                            1.000000   
25%              NaN    999.000000                            7.000000   
50%              NaN   1999.000000                            9.000000   
75%              NaN   1999.000000                           10.000000   
max              NaN   1999.000000                           12.000000   

       maxEstimatedTransferDurationInDays  normalizedExteriorColor  \
count                        51299.000000                      0.0   
mean                            18.868867                      NaN   
std                              5.471971                      NaN   
min                              6.000000                      NaN   
25%                             17.000000                      NaN   
50%                             21.000000                      NaN   
75%                             22.000000                      NaN   
max                             26.000000                      NaN   

       normalizedInteriorColor  review  numberOfFavorites       mpgCity  \
count                      0.0     0.0            87118.0  83641.000000   
mean                       NaN     NaN                0.0     23.493932   
std                        NaN     NaN                0.0      5.962185   
min                        NaN     NaN                0.0      8.000000   
25%                        NaN     NaN                0.0     19.000000   
50%                        NaN     NaN                0.0     23.000000   
75%                        NaN     NaN                0.0     27.000000   
max                        NaN     NaN                0.0    129.000000   

         mpgHighway  originalBatteryRangeInMiles    cylinders  fuelType  \
count  83644.000000                  3216.000000  87118.00000       0.0   
mean      30.077352                   186.421331      4.59269       NaN   
std        5.798370                   126.225841      1.41678       NaN   
min       13.000000                     9.000000      0.00000       NaN   
25%       26.000000                    32.000000      4.00000       NaN   
50%       30.000000                   247.000000      4.00000       NaN   
75%       34.000000                   303.000000      6.00000       NaN   
max      106.000000                   412.000000     12.00000       NaN   

      

Based on the descriptive statistics listed above, cross-referenced with the dataframe head before it, it is safe to exclude the following data from the analysis due to an overwhelming presence of NaN values:

- repairPalData
- normalizedExteriorColor
- normalizedInteriorColor
- review
- fuelType
- recommendationType
- bestFinanceDecision
- vehicleSize
- types
- heroThumbnailImageUrl


The following feature can be removed as well as it provides no useful information to the task at hand:

- numberOfFavorites
- heroImageUrl

In [27]:
# Create a new dataframe that is a copy of the original dataframe
df_copy = combined_df.copy()

# Remove irrelevant columns listed above
columns_to_drop = [
    'repairPalData', 'normalizedExteriorColor', 
    'normalizedInteriorColor', 'review',
    'fuelType', 'recommendationType', 
    'bestFinanceDecision', 'vehicleSize', 
    'types', 'heroThumbnailImageUrl',
    'numberOfFavorites', 'heroImageUrl'
]

df_copy = df_copy.drop(columns = columns_to_drop, axis = 1)
df_copy.head()

stockNumber                vin  year        make              model  \
0     27461478  5NPEG4JA8NH134099  2022     Hyundai             Sonata   
1     26070459  JTDEPMAE5NJ210260  2022      Toyota            Corolla   
2     27175122  3GCUDFE80RG128481  2024   Chevrolet     Silverado 1500   
3     27072297  1V2WE2CA5PC229115  2023  Volkswagen  Atlas Cross Sport   
4     27386026  58AGZ1B12MU110647  2021       Lexus             ES 350   

               body           trim  basePrice  originalPrice  hasPriceDrop  \
0          4D Sedan             SE      19998            NaN         False   
1          4D Sedan             LE      20998            NaN         False   
2       4D Crew Cab  LT Trail Boss      51998            NaN         False   
3  4D Sport Utility      SE w/Tech      32998            NaN         False   
4          4D Sedan        F-Sport      35998            NaN         False   

   mileage  storeId storeName storeCity    state stateAbbreviation  distance  \
0    56782     7298   Roswell   Roswell  Georgia                GA    2108.0   
1    33730     7298   Roswell   Roswell  Georgia                GA    2108.0   
2    58240     7298   Roswell   Roswell  Georgia                GA    2108.0   
3    15895     7298   Roswell   Roswell  Georgia                GA    2108.0   
4    26900     7298   Roswell   Roswell  Georgia                GA    2108.0   

   averageRating  numberOfReviews  isNewArrival  isTransferable  \
0       4.200000                5         False            True   
1       4.500000               22         False            True   
2       5.000000                1         False            True   
3       5.000000                2         False            True   
4       4.333333                6         False            True   

                                            features  \
0  [Rear Defroster, Overhead Airbags, Auxiliary A...   
1  [Power Windows, Power Locks, Air Conditioning,...   
2  [Power Locks, Auxiliary Audio Input, AM/FM Ste...   
3  [Alloy Wheels, 20 Inch Plus Wheels, Power Lock...   
4  [Power Windows, Traction Control, ABS Brakes, ...   

                                 highlightedFeatures  \
0  Rear View Camera,Auxiliary Audio Input,Alloy W...   
1  Rear View Camera,Auxiliary Audio Input,Overhea...   
2  Z71 Package,4WD/AWD,Leather Seats,BOSE Sound S...   
3  Technology Package,Leatherette Seats,Satellite...   
4  Leatherette Seats,Pioneer Sound System,Parking...   

                                          highlights  \
0  [singleOwner, warranty, fuelEfficient, advance...   
1        [warranty, fuelEfficient, advancedFeatures]   
2      [singleOwner, advancedFeatures, premiumAudio]   
3  [singleOwner, lowMiles, warranty, advancedFeat...   
4  [singleOwner, lowMilesPerYear, warranty, advan...   

       lastMadeSaleableDate  transferFee            transferTags  \
0  2025-06-12T20:56:13.565Z       1999.0  [GeneratedByCps, Paid]   
1  2025-06-12T20:53:33.311Z       1999.0  [GeneratedByCps, Paid]   
2  2025-06-20T21:39:32.178Z       1999.0  [GeneratedByCps, Paid]   
3  2025-05-28T20:50:22.353Z       1999.0  [GeneratedByCps, Paid]   
4   2025-06-13T20:59:37.45Z       1999.0  [GeneratedByCps, Paid]   

     transferText   transferType  minEstimatedTransferDurationInDays  \
0  $1999 Shipping  shipping-from                                 9.0   
1  $1999 Shipping  shipping-from                                 9.0   
2  $1999 Shipping  shipping-from                                 9.0   
3  $1999 Shipping  shipping-from                                 9.0   
4  $1999 Shipping  shipping-from                                 9.0   

   maxEstimatedTransferDurationInDays  transferTimesUnavailable exteriorColor  \
0                                22.0                     False          Gray   
1                                22.0                     False          Blue   
2                                22.0                     False         Black   
3                           

In [35]:
# Check Dataframe shape
print("DataFrame Shape:", df_copy.shape)

# Check for missing values
df_copy.isnull().sum()

DataFrame Shape: (87118, 57)


stockNumber                               0
vin                                       0
year                                      0
make                                      0
model                                     0
body                                      2
trim                                   7444
basePrice                                 0
originalPrice                         81118
hasPriceDrop                              0
mileage                                   0
storeId                                   0
storeName                                 0
storeCity                                 0
state                                     0
stateAbbreviation                         0
distance                                  0
averageRating                             0
numberOfReviews                           0
isNewArrival                              0
isTransferable                            0
features                                  0
highlightedFeatures             

Based on the null value counts listed above, the following features can be removed from analysis due to a mjority of the data missing:

- originalPrice (roughly 90% of the data is missing)
- series (almost 90% of the data is missing)
- originalBatteryRangeInMiles (over 90% of the data is missing, probably because most of the vehicles are not electric/hybrid cars)

The following columns can also be removed due to a combinationg of having a relatively high amount of missing values and the fact that transfer fees are generally associated with vehicle transportation between different CarMax storefronts rather than the market price associated with the vehicle itself:

- minEstimatedTransferDurationInDays (about 40% of the data is missing)
- maxEstimatedTransferDurationInDays (about 40% of the data is missing)
- transferFee (about 40% of the data is missing)


In [36]:
# Remove irrelevant columns listed above
columns_to_drop_2 = [
    'originalPrice', 'originalBatteryRangeInMiles', 
    'series', 'minEstimatedTransferDurationInDays',
    'maxEstimatedTransferDurationInDays', 'transferFee'
]

df_copy = df_copy.drop(columns = columns_to_drop_2, axis = 1)
df_copy.head()

stockNumber                vin  year        make              model  \
0     27461478  5NPEG4JA8NH134099  2022     Hyundai             Sonata   
1     26070459  JTDEPMAE5NJ210260  2022      Toyota            Corolla   
2     27175122  3GCUDFE80RG128481  2024   Chevrolet     Silverado 1500   
3     27072297  1V2WE2CA5PC229115  2023  Volkswagen  Atlas Cross Sport   
4     27386026  58AGZ1B12MU110647  2021       Lexus             ES 350   

               body           trim  basePrice  hasPriceDrop  mileage  storeId  \
0          4D Sedan             SE      19998         False    56782     7298   
1          4D Sedan             LE      20998         False    33730     7298   
2       4D Crew Cab  LT Trail Boss      51998         False    58240     7298   
3  4D Sport Utility      SE w/Tech      32998         False    15895     7298   
4          4D Sedan        F-Sport      35998         False    26900     7298   

  storeName storeCity    state stateAbbreviation  distance  averageRating  \
0   Roswell   Roswell  Georgia                GA    2108.0       4.200000   
1   Roswell   Roswell  Georgia                GA    2108.0       4.500000   
2   Roswell   Roswell  Georgia                GA    2108.0       5.000000   
3   Roswell   Roswell  Georgia                GA    2108.0       5.000000   
4   Roswell   Roswell  Georgia                GA    2108.0       4.333333   

   numberOfReviews  isNewArrival  isTransferable  \
0                5         False            True   
1               22         False            True   
2                1         False            True   
3                2         False            True   
4                6         False            True   

                                            features  \
0  [Rear Defroster, Overhead Airbags, Auxiliary A...   
1  [Power Windows, Power Locks, Air Conditioning,...   
2  [Power Locks, Auxiliary Audio Input, AM/FM Ste...   
3  [Alloy Wheels, 20 Inch Plus Wheels, Power Lock...   
4  [Power Windows, Traction Control, ABS Brakes, ...   

                                 highlightedFeatures  \
0  Rear View Camera,Auxiliary Audio Input,Alloy W...   
1  Rear View Camera,Auxiliary Audio Input,Overhea...   
2  Z71 Package,4WD/AWD,Leather Seats,BOSE Sound S...   
3  Technology Package,Leatherette Seats,Satellite...   
4  Leatherette Seats,Pioneer Sound System,Parking...   

                                          highlights  \
0  [singleOwner, warranty, fuelEfficient, advance...   
1        [warranty, fuelEfficient, advancedFeatures]   
2      [singleOwner, advancedFeatures, premiumAudio]   
3  [singleOwner, lowMiles, warranty, advancedFeat...   
4  [singleOwner, lowMilesPerYear, warranty, advan...   

       lastMadeSaleableDate            transferTags    transferText  \
0  2025-06-12T20:56:13.565Z  [GeneratedByCps, Paid]  $1999 Shipping   
1  2025-06-12T20:53:33.311Z  [GeneratedByCps, Paid]  $1999 Shipping   
2  2025-06-20T21:39:32.178Z  [GeneratedByCps, Paid]  $1999 Shipping   
3  2025-05-28T20:50:22.353Z  [GeneratedByCps, Paid]  $1999 Shipping   
4   2025-06-13T20:59:37.45Z  [GeneratedByCps, Paid]  $1999 Shipping   

    transferType  transferTimesUnavailable exteriorColor interiorColor  \
0  shipping-from                     False          Gray          Gray   
1  shipping-from                     False          Blue         Black   
2  shipping-from                     False         Black         Black   
3  shipping-from                     False         Black         Black   
4  shipping-from                     False         Black           Red   

  transmission  isSaleable  isComingSoon  isReserved  mpgCity  mpgHighway  \
0    Automatic        True         False       False     28.0        38.0   
1    Automatic        True         False       False     30.0        38.0   
2    Automatic        True         False       False     23.0        27.0   
3    Automatic        True         False       False     18.0        24.0   
4    Automatic        True 

In [37]:
print("DataFrame Shape:", df_copy.shape)


DataFrame Shape: (87118, 51)


Still need to figure out how to deal with remaining null values in the following columns:

- body                                      2
- trim                                   7444
- lastMadeSaleableDate                   6783
- transferText                              1
- transferType                              1
- exteriorColor                            55
- interiorColor                            83
- transmission                              1
- mpgCity                                3477
- mpgHighway                             3474
- driveTrain                              266
- engineType                                2
- horsepower                             4428
- horsepowerRpm                          5142
- engineSize                               11
- engineTorque                           4606
- engineTorqueRpm                        5919
- featureScore                           8843


In [50]:
# Check rows with duplicate VINS and drop
df_unique = df_copy.drop_duplicates(subset=['vin'], keep='first')
# print("Dataframe with duplicates removed: ")
# print(df_unique)

# Find rows with duplicate VINS
# all_duplicate_vins = df_copy[df_copy.duplicated(subset=['vin'], keep=False)]
# print("\nAll rows with duplicate VIN values:")
# print(all_duplicate_vins)

print("Unique DataFrame Shape:", df_unique.shape)

Dataframe with duplicates removed: 
Unique DataFrame Shape: (86161, 51)


957 Duplicate Rows removed